### ISSUE TITLE
Repeated bases in the adapter sequence with only one mismatch

### ISSUE THREAD

**Issue by AlicePsyche** [marcelm/cutadapt#216]
Hi, 

Thanks for your great tools ~

I read the documentation, it seems that cutadapt deal with mismatch with error rate. I am wondering if I can exactly specify only one mismatch no matter what length the adapter is? 

Actually my reads look like this:
```
GGGGGGGGGGGATTCTTTCATTCCATTATTAATTAATTCATTCAACCATCCATTCTATTATTTATTATTTTATTCAATCATTAATTTATTTTATTATTTTA
GGGGGGGGGCCCTTTTGTCTAATTTATACTACTTACGAAATCAACTCCCACTAACACCTCATCTACATTTACGTCTAACGTCTTCTACCAGATCGTAAAAG
GGGGGGGGGATTCCAACAACAACAACAAAATCAACTTTCATTTTAAAATAATAATTATATTATAAAATAGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT
GGGGGGGGGTAATACTAATATTATCTAATAACACCTCACCATAATCTCATTTACAATTATACATACTCATACTTAATCTAATATTTTTATAATCCGAATAT
GCGGGGGGGTTAACTATATACCTTCTACACCAAATTTTATTTCACTTCACATCAAACCACCACTTTAATTTTAAAACTACCACTTAATATTAACACTTAAG
GGGGGGGGGGAAAACGAATTGGTGGTTAGTTTAGTAGATCGGAAGAGCGTCGTGTAGGGAAAGGGGGTAGATCTCGGGGGTCGCCGGATCATTAAAAAAAA
GGGGGGGGGGGGGTGCCAAAACCACCCAAAAAAAAGACGGGACAAGCAACGAGAAGGAAAGGCGGGCAGATCCCACGCGCTGCCTTTACAATAAAAAAAAA
GGGGGGGGGAATAATTAAAAAAATAAAATAAAAACTAGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTAAAAAAA
GGGGGGGGGGTTCCTATATAACTCCCAATAATACTTATAAATTTCTTCCCACCCCAACCTTGATTTCGCTTTTAAAAACGATACCACGTAAAAACAATCTA
GGGGGGGGGGGCTAAATTATTAACCCTTATTTTTTTTTATTTTTTCCCCAATTTCTATTTAACGAAAAATATTTTAACACATTTCTAAACAATATAATTTT
GGGGGGGGGGCGCAATGAGTCTCCCTTTTTTGTGATCAGCGACTTCCCCTACATGACCATGCCCAGTGGGCCGTCTAAGATCGGAAGCGCGTCGTGTCGGG
```

As you can see, the 5' ployG number is different, ranging from 6-30, I used to remove the polyG adapter using regular expression `^G{1,}[AGCTN]{0,1}G{1,}`, so the adapter `GCGGGGGGG` could also be removed. How can I achieve this by cutadapt?  

Thanks in advance!

**Comment by marcelm** [comment]
Cutadapt always requires that you specify an error rate; you cannot specify an absolute number of errors. I think this is a good thing when the length of the adapter match is variable: Few errors are allowed in shorter matches, more errors in longer matches. The only negative thing is that you sometimes need to calculate what this means.

Cutadapt by default uses an error rate of 0.1 (that is, 10%). The actual number of errors allowed depends on the length of the match. For a match of length 10, it is 10 bp * 0.1 errors/bp = 1 error. In the example you have above with `GCGGGGGGG`, the match has a length of 9 bp, which gives 9 * 0.1 = 0.9, which is rounded down to 0 allowed errors. This is why it is not trimmed: For any match shorter than 10 bp, no mismatches are allowed at all. Cutadapt tells you this in the summary at the end: "No. of allowed errors: 0-9 bp: 0; 10-16 bp: 1".

You should in my opinion just increase the error rate, perhaps by a large amount. If you want to allow 1 out of 9 mismatches, specify `-e 0.12` (1/9=0.111..., then rounded up). Or for 1 out of 8 mismatches, specify `-e 0.13` (1/8=0.125, then rounded up), and so on. I haven’t tested it in detail, but you can probably go up quite a lot, possibly up to 0.25 or 0.34 (for one mismatch per three or four bases, respectively), without getting too many false positive matches.

There will always be artifacts, by the way. If a read starts with `GAGCCAATT...`, for example, then from the read alone one cannot decide whether there is a poly-G with one mismatch or whether the `GAG` is just part of normal read.

In summary, I would try a variation of this command:
```
cutadapt -g G{40} -e 0.25 -o output.fastq.gz input.fastq.gz
```
The `G{40}` is an abbreviation for fourty times `G`, and `-e 0.25` is the error rate.

**Comment by AlicePsyche** [comment]
Thanks for your quick and helpful response! I would test this immediately ~~~

**Comment by marcelm** [comment]
I forgot to close this, will do so now. Note that I think I will add the possibility of specifying an absolute number of errors in the future, see #358.
